core/src/bytecode/value/mod.rs

```diff
@@ -427,6 +435,138 @@ impl NickelValue {
         }
     }
 
+    /// Returns the body tag of the underlying value block, or `None` if `self` is an inline value.
+    pub fn body_tag(&self) -> Option<BodyTag> {
+        (matches!(self.tag(), ValueTag::Pointer)).then(|| {
+            // Safety: if `self.tag()` is `Pointer`, then `self.data` must be valid pointer to a
+            // value  block.
+            unsafe {
+                let as_ptr = NonNull::new_unchecked(self.data as *mut u8);
+                ValueBlockRc::tag_from_raw(as_ptr)
+            }
+        })
+    }
+
+    /// Determines if a value is in evaluated form, called weak head normal form (WHNF). See
+    /// [crate::term::Term::is_whnf] for more detais.
+    pub fn is_whnf(&self) -> bool {
+        matches!(self.body_tag(), Some(BodyTag::Thunk | BodyTag::Term))
```

Comment:
Is there a missing negation here?

Reply:
Argh, there is. Good catch, thanks! That could have been quite painful to debug.